### PR TITLE
#14696 Generate the settings object automatically based on the capabi…

### PR DESCRIPTION
**Pull request by vtkalek**
…lities file

#14658 Improve new visual template in pbiviz tool

### CODE REVIEW

File: lib/VisualGenerator.js [hunk 1]
```diff
@@ -89,7 +89,102 @@ function copyVisualTemplate(targetPath, templateName) {
...
+                                console.log('fill');
+                            }
+                            settingsSubClassDefinition += ': string = "";\n';
+                        } else
+                        if (propertyType.formatting) {
+                            if (verbose) {
+                                console.log('formatting');
+                            }
+                            if (propertyType.formatting.fontSize)
+                            settingsSubClassDefinition += ': number = 12;\n';
```
Comment (savvinsergey) [inline lib/VisualGenerator.js]: Please add '{}' to here

File: spec/helpers/FileSystem.js [hunk 2]
```diff
@@ -86,7 +86,22 @@ class FileSystem {
...
+     * Executes console commands
+     * 
+     * @param {string} command - the command to be executed
+     * @param {string} pathToExec - path where to execute command
+     * @param {string} pathToReturn - path to return after command execution
+     */
+    static runCMDCommand(command, pathToExec, pathToReturn) {
+        const execSync = require('child_process').execSync;
+        process.chdir(pathToExec);
+        let cmd2 = execSync('npm i --silent');
```
Comment (AviSander) [inline spec/helpers/FileSystem.js]: cmd2 needed?

File: package.json [hunk 3]
```diff
@@ -38,14 +38,16 @@
     "typescript": "2.1.5",
     "uglify-js": "^2.6.2",
     "uuid": "^3.0.0",
-    "gulp-powerbi-package-validator": "^1.0.0"
+    "gulp-powerbi-package-validator": "^1.0.0",
+    "powerbi-visuals-utils-dataviewutils": "1.0.1"
```
Comment (AviSander) [inline package.json]: is this required both here and in devDependencies?

File: spec/e2e/pbivizNewSpec.js [hunk 4]
```diff
@@ -110,7 +110,8 @@ describe("E2E - pbiviz new", () => {
                 visualPath = path.join(tempPath, visualName);
 
             FileSystem.runPbiviz('new', visualName, `--template ${template}`);
```
Comment (AviSander) [inline spec/e2e/pbivizNewSpec.js]: would it make sense to have 'pbiviz new'  also run 'npm install' now?

File: lib/VisualGenerator.js [hunk 5]
```diff
@@ -142,6 +239,7 @@ class VisualGenerator {
                     }
 
                     copyVisualTemplate(visualPath, buildOptions.template);
+                    generateSettings(visualPath);
```
Comment (AviSander) [inline lib/VisualGenerator.js]: We should think of how to do an 'update settings' option for the user. Meaning that once the user goes and updates his capabilities.json he can then ask to just update his settings.ts file.
I think we can do this in a separate PR \ iteration.

File: lib/VisualGenerator.js [hunk 6]
```diff
@@ -89,7 +89,104 @@ function copyVisualTemplate(targetPath, templateName) {
...
+function generateSettings(targetPath, verbose) {
+    let capabilitiesFileName = targetPath + '/capabilities.json';
+
+    fs.readJson(capabilitiesFileName, (err, packageObj) => {
+        if (err) {
+            console.error(err);
+            return;
+        }
+        if (verbose) {
+            console.log('parce capabilities.json. Searching for settings');
```
Comment (AviSander) [inline lib/VisualGenerator.js]: parse

File: lib/VisualGenerator.js [hunk 7]
```diff
@@ -89,7 +89,104 @@ function copyVisualTemplate(targetPath, templateName) {
...
+            }
+            if (verbose) {
+                console.log('final text: ' + VisualSettingsMethods);
+            }
+
+        } else {
+            if (verbose) {
+                console.log('Settings not found');
+            }
+            VisualSettingsMethods += '// TODO: fill all visual settings here\n ' +
```
Comment (AviSander) [inline lib/VisualGenerator.js]: this should also be on our list: default values directly from capabilities.
Also in separate iteration

File: lib/VisualGenerator.js [hunk 8]
```diff
@@ -89,7 +89,104 @@ function copyVisualTemplate(targetPath, templateName) {
     fs.copySync(path.join(VISUAL_TEMPLATES_PATH, '_global'), targetPath);
     fs.copySync(path.join(VISUAL_TEMPLATES_PATH, templateName), targetPath);
 }
+/**
```
Comment (AviSander) [inline lib/VisualGenerator.js]: does the capabilites.json pass schema validation before the function is called?

File: lib/VisualGenerator.js [hunk 9]
```diff
@@ -89,7 +89,104 @@ function copyVisualTemplate(targetPath, templateName) {
...
+ * Generate settings.ts
+ * 
+ * @param {string} targetPath - file path to root of visual package
+ * @param {string} verbose - to verbose output
+ */
+function generateSettings(targetPath, verbose) {
```
Comment (AviSander) [inline lib/VisualGenerator.js]: There should be a UT that tests the VisualSettings object. 
- make sure the different members can be access (run)
- use templates/visuals/default/capabilities.json to get a good coverage
Since we're generating code we want to ensure it is 'runnable'. Test should run in pbiviz tools tests - not in the generated visual tests.

File: lib/VisualGenerator.js (hunk 8 above)
Comment (vtkalek) [inline lib/VisualGenerator.js]: There is a call to validTemplate function before it.  

 if (!validTemplate(buildOptions.template)) {
                return reject(new Error('Invalid template'));
            }

File: spec/e2e/pbivizNewSpec.js (hunk 4 above)
Comment (vtkalek) [inline spec/e2e/pbivizNewSpec.js]: I'm not sure about it. I think we should left it for user - to run npm install command

File: lib/VisualGenerator.js (hunk 9 above)
Comment (vtkalek) [inline lib/VisualGenerator.js]: Added UT to check if settings.ts file generated correctly
